### PR TITLE
Filter

### PR DESCRIPTION
**Pull request by tenadopt**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -21,14 +21,14 @@
     "@hookform/resolvers": "^3.9.0",
     "@mui/material": "^5.16.0",
     "@reduxjs/toolkit": "^2.2.6",
+    "@types/react-router-dom": "^5.3.3",
```
Comment (shev7) [inline package.json]: should be dev dependecy

File: src/app/routes.tsx [hunk 2]
```diff
@@ -1,67 +1,115 @@
...
-const HomePage = lazy(()=>import('pages/Home/ui/HomePage'));
-const SignInPage = lazy(()=>import('pages/SignIn/ui/SignInPage'));
-const SignUpPage = lazy(()=>import('pages/SignUp/ui/SignUpPage'));
-const ImageFeedPage = lazy(()=>import('pages/ImageFeed/ui/ImageFeedPage'));
-const ImageEditorPage = lazy(()=>import('pages/ImageEditor/ui/ImageEditorPage'));
+interface RequireAuthProps {
+    children: ReactNode;
+}
+
+const RequireAuth: React.FC<RequireAuthProps> = ({ children }) => {
```
Comment (shev7) [inline src/app/routes.tsx]: no need to use to type components
```suggestion
const RequireAuth = ({ children }: RequireAuthProps) => {
```

File: src/app/routes.tsx [hunk 3]
```diff
@@ -1,67 +1,115 @@
...
+};
+
+const PublicLayout: React.FC = () => {
+    const auth = useAppSelector((state) => state.auth);
+
+    return auth.userId ? <Navigate to="/feed" replace /> : (
+        <AppLayout>
+            <Outlet />
+        </AppLayout>
+    );
```
Comment (shev7) [inline src/app/routes.tsx]:
```suggestion
if (auth.userId) {
   return <Navigate to="/feed" replace />
}

    return (
        <AppLayout>
            <Outlet />
        </AppLayout>
    );
```
please simplify conditions, avoid using ternary operators

File: src/app/routes.tsx [hunk 4]
```diff
@@ -1,67 +1,115 @@
...
-        element: <AppLayout />,
+        path: '/',
+        element: <PublicLayout />,
         errorElement: <ErrorBoundary />,
         children: [
             {
-                path: "",
+                path: '',
                 element: (
                     <Suspense fallback={<Box display="flex" justifyContent="center" alignItems="center" height="100vh"><CircularProgress /></Box>}>
```
Comment (shev7) [inline src/app/routes.tsx]: use suspense inside of layout

Comment (shev7) [reply src/app/routes.tsx]: in this kase you want need to wrap each page into it

File: src/features/imageGallery/api/imageService.ts [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+    const q = userId
+        ? query(imagesCollection, where('userId', '==', userId), orderBy('createdAt', 'desc'))
+        : query(imagesCollection, orderBy('createdAt', 'desc'));
+
+    const querySnapshot = await getDocs(q);
+    return querySnapshot.docs.map(doc => ({
+        id: doc.id,
+        ...doc.data(),
+        createdAt: doc.data().createdAt.toDate()
+    })) as Image[];
```
Comment (shev7) [inline src/features/imageGallery/api/imageService.ts]:
```suggestion
    return querySnapshot.docs.map<Image>(doc => ({
        id: doc.id,
        ...doc.data(),
        createdAt: doc.data().createdAt.toDate()
    }));
```

File: src/features/works/api/worksServices.ts [hunk 6]
```diff
@@ -0,0 +1,41 @@
+import { collection, doc, getDoc, setDoc, addDoc, deleteDoc, query, where, getDocs } from 'firebase/firestore';
+import { db } from 'firebaseConfig';
+import { Work } from 'features/works/types';
+
+export const fetchUserWorks = async (userId: string): Promise<Work[]> => {
+    const worksCollection = collection(db, 'works');
+    const q = query(worksCollection, where('userId', '==', userId));
+    const querySnapshot = await getDocs(q);
+    return querySnapshot.docs.map((doc) => ({
```
Comment (shev7) [inline src/features/works/api/worksServices.ts:9]:
```suggestion
    const querySnapshot = await getDocs(q);
    
    return querySnapshot.docs.map((doc) => ({
```

File: src/features/works/api/worksServices.ts [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+};
+
+export const fetchWorkById = async (id: string): Promise<Work | null> => {
+    const docRef = doc(db, 'works', id);
+    const docSnap = await getDoc(docRef);
+    if (docSnap.exists()) {
+        return { id: docSnap.id, ...(docSnap.data() as Omit<Work, 'id'>), createdAt: docSnap.data().createdAt.toDate() };
+    } else {
+        return null;
+    }
```
Comment (shev7) [inline src/features/works/api/worksServices.ts]:
```suggestion
    if (docSnap.exists()) {
        return { id: docSnap.id, ...(docSnap.data() as Omit<Work, 'id'>), createdAt: docSnap.data().createdAt.toDate() };

        return null;
```

or more simplified

```suggestion
    if (!docSnap.exists()) {
        return null;
        };

       return { id: docSnap.id, ...(docSnap.data() as Omit<Work, 'id'>), createdAt: docSnap.data().createdAt.toDate() 
```

File: src/features/works/api/worksServices.ts [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+};
+
+export const addWork = async (work: Omit<Work, 'id' | 'createdAt'>): Promise<void> => {
+    await addDoc(collection(db, 'works'), {
+        ...work,
+        createdAt: new Date()
+    });
+};
```
Comment (shev7) [inline src/features/works/api/worksServices.ts]:
```suggestion
export const addWork = (work: Omit<Work, 'id' | 'createdAt'>): Promise<void> => {
   return addDoc(collection(db, 'works'), {
        ...work,
        createdAt: new Date()
    });
};
```
redundant async/await, await it where it's needed

File: src/features/works/api/worksServices.ts [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+
+export const updateWork = async (id: string, updatedData: Partial<Omit<Work, 'id' | 'createdAt'>>): Promise<void> => {
+    const docRef = doc(db, 'works', id);
+    await setDoc(docRef, { ...updatedData, updatedAt: new Date() }, { merge: true });
+};
+
+export const deleteWork = async (id: string): Promise<void> => {
+    const docRef = doc(db, 'works', id);
+    await deleteDoc(docRef);
+};
```
Comment (shev7) [inline src/features/works/api/worksServices.ts]: same here with async/await, just return like in previous example

File: src/pages/Home/ui/HomePage.tsx [hunk 10]
```diff
@@ -1,89 +1,129 @@
...
-                setError('Failed to load works')
+                setError('Failed to load works');
             } finally {
                 setLoading(false);
             }
         };
 
-        loadWorks();
+        // Call loadWorks and handle the promise
+        loadWorks().catch(console.error);
```
Comment (shev7) [inline src/pages/Home/ui/HomePage.tsx]: add notifications, for example using react-toastify

File: src/pages/Home/ui/HomePage.tsx [hunk 11]
```diff
@@ -1,89 +1,129 @@
...
+                                    <Typography variant="h6">{work.title}</Typography>
+                                    <Typography variant="body2">{work.description}</Typography>
+                                    <Button
+                                        onClick={(e) => {
+                                            e.stopPropagation();
+                                            handleOpenDialog(work.id);
+                                        }}
+                                        color="secondary"
+                                    >
+                                        Delete
```
Comment (shev7) [inline src/pages/Home/ui/HomePage.tsx]: this is very hard to read due to ternaries, please avoid using them

File: src/pages/ImageEditor/ui/ImageEditorPage.tsx [hunk 12]
```diff
@@ -1,11 +1,146 @@
...
+                } catch (err) {
+                    setError('Failed to load work');
+                } finally {
+                    setLoading(false);
+                }
+            } else {
+                setLoading(false);
+            }
+        };
+        loadWork().catch(console.error);
```
Comment (shev7) [inline src/pages/ImageEditor/ui/ImageEditorPage.tsx]: really hard to read condition, avoid if blocks nesting, try to use early returns like we discussed

File: src/pages/ImageEditor/ui/ImageEditorPage.tsx [hunk 13]
```diff
@@ -1,11 +1,146 @@
...
+                            error={!!errors.title}
+                            helperText={errors.title?.message}
+                        />
+                        <TextField
+                            label="Description"
+                            fullWidth
+                            margin="normal"
+                            {...register('description' as FieldPath<WorkFormValues>)}
+                            error={!!errors.description}
+                            helperText={errors.description?.message}
```
Comment (shev7) [inline src/pages/ImageEditor/ui/ImageEditorPage.tsx]: also unreadable due to ternaries

File: src/pages/ImageEditor/ui/ImageEditorPage.tsx [hunk 14]
```diff
@@ -1,11 +1,146 @@
...
+                            label="Image URL"
+                            fullWidth
+                            margin="normal"
+                            {...register('imageUrl'  as FieldPath<WorkFormValues>)}
+                            error={!!errors.imageUrl}
+                            helperText={errors.imageUrl?.message}
+                        />
+                        <CanvasEditor imageUrl={initialData?.imageUrl || ''} />
+                        <Button variant="contained" color="primary" type="submit" disabled={loading}>
+                            {loading ? 'Saving...' : isEditing ? 'Update Work' : 'Create Work'}
```
Comment (shev7) [inline src/pages/ImageEditor/ui/ImageEditorPage.tsx]: spinner component can be used

File: src/pages/ImageFeed/ui/ImageFeedPage.tsx [hunk 15]
```diff
@@ -1,11 +1,75 @@
...
+                                />
+                                <CardContent>
+                                    <Typography variant="body2" color="textSecondary">
+                                        Uploaded by: {image.userId}
+                                    </Typography>
+                                    <Typography variant="body2" color="textSecondary">
+                                        {new Date(image.createdAt).toLocaleString()}
+                                    </Typography>
+                                </CardContent>
+                            </Card>
```
Comment (shev7) [inline src/pages/ImageFeed/ui/ImageFeedPage.tsx]: again ternaries

File: src/pages/Work/ui/WorkPage.tsx [hunk 16]
```diff
@@ -0,0 +1,68 @@
...
+
+    const handleUpdate = async () => {
+        if (!work || !workId) return;
+        const updatedData = { title: 'New Title' }; // Example of updated data
+        try {
+            await updateWork(workId, updatedData);
+            setWork({ ...work, ...updatedData });
+            alert('Work updated successfully');
+        } catch (error) {
+            console.error('Error updating work:', error);
```
Comment (shev7) [inline src/pages/Work/ui/WorkPage.tsx:51]: notification

File: src/pages/Work/ui/WorkPage.tsx [hunk 17]
```diff
@@ -0,0 +1,68 @@
...
+                <Box display="flex" justifyContent="center" mt={2}>
+                    <Typography color="error">{error}</Typography>
+                </Box>
+            ) : work ? (
+                <Box>
+                    <Typography variant="h4">{work.title}</Typography>
+                    <Typography variant="body1">{work.description}</Typography>
+                    <Button onClick={handleUpdate} color="primary">Update Work</Button>
+                </Box>
+            ) : null}
```
Comment (shev7) [inline src/pages/Work/ui/WorkPage.tsx]: ternaries

File: src/widgets/AppLayout.tsx [hunk 18]
```diff
@@ -2,16 +2,16 @@ import React from 'react';
...
-import { Container } from '@mui/material';
+import {Box, Container} from '@mui/material';
 
 const AppLayout = () => (
-    <>
+    <Box display="flex" flexDirection="column" minHeight="100vh">
         <Header />
-        <Container component="main" maxWidth="md">
+        <Container component="main" maxWidth="md" style={{ flexGrow: 1, paddingTop: '20px', paddingBottom: '20px' }}>
             <Outlet />
```
Comment (shev7) [inline src/widgets/AppLayout.tsx]: you should apply style using sx prop

**Review by shev7 — CHANGES_REQUESTED** [review]

File: package.json (hunk 1 above)
Comment (tenadopt) [reply package.json]: Done

File: src/app/routes.tsx (hunk 2 above)
Comment (tenadopt) [reply src/app/routes.tsx]: Done

File: src/features/imageGallery/api/imageService.ts (hunk 5 above)
Comment (tenadopt) [reply src/features/imageGallery/api/imageService.ts]: If i write that i have an ERROR 
Argument type (doc: any) => {[p: string]: any, createdAt: Date, id: string} is not assignable to parameter type (value: QueryDocumentSnapshot<DocumentData, DocumentData>, index: number, array: QueryDocumentSnapshot<DocumentData, DocumentData>[]) => Image

File: src/features/works/api/worksServices.ts (hunk 7 above)
Comment (tenadopt) [reply src/features/works/api/worksServices.ts]: Done

File: src/features/works/api/worksServices.ts (hunk 8 above)
Comment (tenadopt) [reply src/features/works/api/worksServices.ts]: If we don't use async we have warning
Promise returned from addDoc is ignored

File: src/features/works/api/worksServices.ts (hunk 9 above)
Comment (tenadopt) [reply src/features/works/api/worksServices.ts]: Promise returned from setDoc is ignored

File: src/features/works/api/worksServices.ts (hunk 6 above)
Comment (tenadopt) [reply src/features/works/api/worksServices.ts:9]: Done

File: src/app/routes.tsx (hunk 3 above)
Comment (tenadopt) [reply src/app/routes.tsx]: Done

File: src/pages/Home/ui/HomePage.tsx (hunk 10 above)
Comment (tenadopt) [reply src/pages/Home/ui/HomePage.tsx]: Done

File: src/pages/Home/ui/HomePage.tsx (hunk 11 above)
Comment (tenadopt) [reply src/pages/Home/ui/HomePage.tsx]: Done

File: src/pages/ImageEditor/ui/ImageEditorPage.tsx (hunk 14 above)
Comment (tenadopt) [reply src/pages/ImageEditor/ui/ImageEditorPage.tsx]: Done

File: src/pages/ImageEditor/ui/ImageEditorPage.tsx (hunk 13 above)
Comment (tenadopt) [reply src/pages/ImageEditor/ui/ImageEditorPage.tsx]: Done

File: src/pages/Work/ui/WorkPage.tsx (hunk 16 above)
Comment (tenadopt) [reply src/pages/Work/ui/WorkPage.tsx:51]: Done

File: src/pages/Work/ui/WorkPage.tsx (hunk 17 above)
Comment (tenadopt) [reply src/pages/Work/ui/WorkPage.tsx]: Done

File: src/features/imageGallery/api/imageService.ts (hunk 5 above)
Comment (tenadopt) [reply src/features/imageGallery/api/imageService.ts]: Done

File: src/pages/ImageFeed/ui/ImageFeedPage.tsx (hunk 15 above)
Comment (tenadopt) [reply src/pages/ImageFeed/ui/ImageFeedPage.tsx]: Done

File: src/app/routes.tsx [hunk 19]
```diff
@@ -1,67 +1,97 @@
...
+        return <Navigate to="/" state={{ from: location }} replace />;
+    }
+
+    return <>{children}</>;
+};
+
+const PublicLayout: React.FC = () => {
+    const auth = useAppSelector((state) => state.auth);
+    if (auth.userId) return <Navigate to="/feed" replace />;
+    return (
```
Comment (shev7) [inline src/app/routes.tsx:29]:
```suggestion
    const auth = useAppSelector((state) => state.auth);
    
    if (auth.userId) return <Navigate to="/feed" replace />;
    
    return (
```
folow code style

File: src/app/routes.tsx
Comment (shev7) [inline src/app/routes.tsx:1]: error boundary should be attached to each element, to each route

File: src/features/works/api/worksServices.ts [hunk 20]
```diff
@@ -0,0 +1,42 @@
...
+        id: doc.id,
+        ...(doc.data() as Omit<Work, 'id'>),
+        createdAt: doc.data().createdAt.toDate()
+    }));
+};
+
+export const fetchWorkById = async (id: string): Promise<Work | null> => {
+    const docRef = doc(db, 'works', id);
+    const docSnap = await getDoc(docRef);
+    if (!docSnap.exists()) {
```
Comment (shev7) [inline src/features/works/api/worksServices.ts:19]:
```suggestion
    const docSnap = await getDoc(docRef);
    
    if (!docSnap.exists()) {
```

File: src/pages/Home/ui/HomePage.tsx [hunk 21]
```diff
@@ -1,91 +1,127 @@
...
-            <Container maxWidth='md'>
-                <Box display="flex" justifyContent="center" alignItems="center" height="100vh">
-                    <CircularProgress/>
-                </Box>
-            </Container>
-        )
-    }
+    const handleDelete = async () => {
+        if (!selectedWorkId) return;
+        try {
```
Comment (shev7) [inline src/pages/Home/ui/HomePage.tsx:56]:
```suggestion
        if (!selectedWorkId) return;
        
        try {
```

File: src/pages/ImageEditor/ui/ImageEditorPage.tsx [hunk 22]
```diff
@@ -1,11 +1,163 @@
...
+
+            } catch (err) {
+                toast.error('Failed to load work');
+                setError('Failed to load work');
+            } finally {
+                setLoading(false);
+            }
+        };
+
+        loadWork().catch(console.error);
```
Comment (shev7) [inline src/pages/ImageEditor/ui/ImageEditorPage.tsx:83]:
```suggestion
        loadWork().catch(console.error);
```
toast

File: src/pages/ImageEditor/ui/ImageEditorPage.tsx [hunk 23]
```diff
@@ -1,11 +1,163 @@
...
+    }, [workId, userId, setValue]);
+
+    const onSubmit: SubmitHandler<WorkFormValues> = async (data) => {
+        if (!userId) return;
+        setLoading(true);
+        try {
```
Comment (shev7) [inline src/pages/ImageEditor/ui/ImageEditorPage.tsx:89]:
```suggestion
        if (!userId) return;
        
        setLoading(true);
        
        try {
```

**Review by shev7 — CHANGES_REQUESTED** [review]

**Pull request merged**